src/rag/system.py

```diff
@@ -0,0 +1,119 @@
+from typing import Protocol, cast
+
+from langchain_core.documents import Document
+from langchain_core.embeddings import Embeddings
+from pydantic import BaseModel
+
+from rag.generator import Generator
+from rag.query_generator import QueryGenerator
+from rag.retriever import HanaDBRetriever
+from utils.hana import create_hana_connection
+from utils.logging import get_logger
+from utils.models.factory import IModel, ModelType
+from utils.settings import (
+    DATABASE_PASSWORD,
+    DATABASE_PORT,
+    DATABASE_URL,
+    DATABASE_USER,
+    DOCS_TABLE_NAME,
+)
+
+logger = get_logger(__name__)
+
+
+class Query(BaseModel):
+    """A RAG system query."""
+
+    text: str
+
+
+class IRAGSystem(Protocol):
+    """A protocol for a RAG system."""
+
+    def search(self, query: Query) -> str:
+        """Query the system with a given query."""
+        ...
+
+    def retrieve(self, query: Query, top_k: int = 5) -> list[Document]:
+        """Retrieve documents for a given query."""
+        ...
+
+    def generate(
+        self,
+        query: Query,
+        relevant_docs: list[Document],
+    ) -> str:
+        """Generate a response to a given query."""
+        ...
+
+
+class RAGSystem:
+    """A system that can be used to generate queries and retrieve documents."""
+
+    def __init__(self, models: dict[str, IModel | Embeddings]):
+        # setup query generator
+        self.query_generator = QueryGenerator(
+            cast(IModel, cast(IModel, models[ModelType.GPT4O_MINI]))
+        )
+        # setup retriever
+        hana_conn = create_hana_connection(
+            DATABASE_URL, DATABASE_PORT, DATABASE_USER, DATABASE_PASSWORD
+        )
```

Comment:
Does it make sense instead of creating the connection here, and have a lock on a specific vector-db impl, to use an abstract connection? Also this will make mocking and testing more easy!

Reply:
I am not sure about this. If you look at the `test_retriever.py` unit tests file. It is mocking with dbapi.Connection interface:
```
@pytest.fixture
def mock_connection():
    """Create a mock HANA DB connection."""
    return Mock(spec=dbapi.Connection)
``` 

Python is different that Go where you can easily mock.